hal/ia32/console.c

```diff
@@ -21,133 +21,343 @@
-		if (console.col == console.maxcol) {
-			console.row++;
-			console.col = 0;
+		/* End of line */
+		if (col == halconsole_common.cols) {
+			row++;
+			col = 0;
 		}
 
-		/* scroll down */
-		if (console.row == console.rows) {
-			hal_memcpy(console.vram, console.vram + console.maxcol, console.maxcol * (console.rows - 1) * 2);
-			console.row--;
-			console.col = 0;
-			hal_memsetw(console.vram + console.maxcol * (console.rows - 1), attr << 8 | ' ', console.maxcol);
+		/* Scroll down */
+		if (row == halconsole_common.rows) {
+			i = halconsole_common.cols * (halconsole_common.rows - 1);
+			console_memmove(halconsole_common.vram, halconsole_common.vram + halconsole_common.cols, i);
+			console_memset(halconsole_common.vram + i, (u16)halconsole_common.attr << 8 | ' ', halconsole_common.cols);
+			row--;
+			col = 0;
 		}
 
-		/* Set hardware cursor position */
-		hal_outb(console.crtc, CRTC_CURSORH);
-		hal_outb(console.crtc + 1, (console.row * console.maxcol + console.col) >> 8);
-		hal_outb(console.crtc, CRTC_CURSORL);
-		hal_outb(console.crtc + 1, (console.row * console.maxcol + console.col) & 0xff);
+		/* Update cursor */
+		i = row * halconsole_common.cols + col;
+		hal_outb(halconsole_common.crtc, 0x0e);
+		hal_outb((void *)((addr_t)halconsole_common.crtc + 1), i >> 8);
+		hal_outb(halconsole_common.crtc, 0x0f);
+		hal_outb((void *)((addr_t)halconsole_common.crtc + 1), i);
+		*((u8 *)(halconsole_common.vram + i) + 1) = halconsole_common.attr;
 	}
-	console.attr = attr;
-	hal_spinlockClear(&console.spinlock, &sc);
 
-	return;
+	hal_spinlockClear(&halconsole_common.spinlock, &sc);
+}
+
+
+void hal_consolePrint(int attr, const char *s)
+{
+	if (attr == ATTR_BOLD)
+		_hal_consolePrint(CONSOLE_BOLD);
+	else if (attr != ATTR_USER)
+		_hal_consolePrint(CONSOLE_CYAN);
+
+	_hal_consolePrint(s);
+	_hal_consolePrint(CONSOLE_NORMAL);
 }
 
 
 void hal_consolePutch(char c)
 {
 	const char str[] = { c, '\0' };
-	hal_consolePrint(console.attr, str);
+	_hal_consolePrint(str);
 }
 
 
 __attribute__ ((section (".init"))) void _hal_consoleInit(void)
 {
-	int isColor = 0;
-
-	/* Test monitor type */
-	isColor = (hal_inb((void *)MAIN_MISCIN) & 0x01);
-
-	console.rows = 25;
-	console.maxcol = 80;
-	console.row = 0;
-	console.col = 0;
-	console.attr = ATTR_NORMAL;
+	unsigned char color;
 
-	console.vram = VADDR_KERNEL + (isColor ? VRAM_COLOR : VRAM_MONO);
-	console.crtc = isColor ? (void *)BASE_COLOR : (void *)BASE_MONO;
+	/* Check color support */
+	color = hal_inb((void *)0x3cc) & 0x01;
 
-	hal_memsetw(console.vram, ' ' | ATTR_NORMAL << 8, console.rows * console.maxcol);
+	/* Initialize VGA */
+	halconsole_common.vram = (u16 *)(VADDR_KERNEL + (color ? 0xb8000 : 0xb0000));
+	halconsole_common.crtc = (void *)(color ? 0x3d4 : 0x3b4);
 
-	hal_spinlockCreate(&console.spinlock, "console.spinlock");
-	return;
+	/* Default 80x25 text mode with magenta color attribute */
+	halconsole_common.rows = 25;
+	halconsole_common.cols = 80;
+	halconsole_common.attr = 0x05;
```

Comment:
In kernel we want to initialize default color attribute to cyan - I'd change the comment and set `halconsole_common.attr = 0x03`

Reply:
Done